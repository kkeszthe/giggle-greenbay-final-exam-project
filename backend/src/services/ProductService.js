import { userService } from '../dependencies/dependencyInjection';

export class ProductService {
  constructor(productRepo, errorCodes) {
    this.product = productRepo;
    this.user = userService;
    this.errorCodes = errorCodes;
  }

  validateURL(string) {
    var pattern = new RegExp(
      '^(https?:\\/\\/)?' + // protocol
        '((([a-z\\d]([a-z\\d-]*[a-z\\d])*)\\.)+[a-z]{2,}|' + // domain name
        '((\\d{1,3}\\.){3}\\d{1,3}))' + // OR ip (v4) address
        '(\\:\\d+)?(\\/[-a-z\\d%_.~+]*)*' + // port and path
        '(\\?[;&a-z\\d%_.~+=-]*)?' + // query string
        '(\\#[-a-z\\d_]*)?$',
      'i'
    ); // fragment locator
    return !!pattern.test(string);
  }

  async getSellableProducts() {
    const products = await this.product.getSellableProducts();
    return { products };
  }
  async getById({ productId }) {
    return (await this.product.getById({ productId }))[0];
  }

  async add({ product_name, description, photo_url, price, seller_id }) {
    if (!product_name || !description || !seller_id)
      throw new Error(this.errorCodes.missingParam);

    if (!price || !(Number.isInteger(price) && price > 0))
      throw new Error(this.errorCodes.invalidPrice);
    if (!photo_url || this.validateURL(photo_url))
      throw new Error(this.errorCodes.invalidURL);
    const productId = (
      await this.product.add({
        product_name,
        description,
        photo_url,
        price,
        seller_id,
      })
    ).insertId;
    return await this.getById({ productId });
  }

  async sellProduct({ productId, userId }) {
    if (!productId || !userId) throw new Error(this.errorCodes.missingParam);
    const product = await this.getById({ productId });
    const buyer = await this.user.getById({ userId });
    if (!product) throw new Error(this.errorCodes.invalidProductId);
    if (product.price > buyer.balance)
      throw new Error(this.errorCodes.invalidBalance);
    if (product.buyer_id) throw new Error(this.errorCodes.usedUserId);
    await this.product.sellProduct({ productId, userId });
    await this.user.updateBalance({ userId, amount: product.price * -1 });
    await this.user.updateBalance({
      userId: product.seller_id,
      amount: product.price,
    });
    return await this.getById({ productId });
  }
}
