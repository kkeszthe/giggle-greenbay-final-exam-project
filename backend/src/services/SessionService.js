import jwt from 'jsonwebtoken';

export class SessionService {
  constructor(userService, errorCodes) {
    this.user = userService;
    this.errorCodes = errorCodes;
  }

  validateParams({ username, password }) {
    if (!username && !password)
      throw new Error(this.errorCodes.missingUsernameAndPassword);
    if (!username) throw new Error(this.errorCodes.missingUsername);
    if (!password) throw new Error(this.errorCodes.missingPassword);
  }

  getToken({ userId }) {
    return jwt.sign({ userId }, process.env.JWT_SECRET || 'secret');
  }

  async login({ username, password }) {
    this.validateParams({ username, password });
    const user = await this.user.getByUsernameAndPassword({
      username,
      password,
    });
    if (user.length === 0)
      throw new Error(this.errorCodes.invalidUsernameAndPassword);
    return {
      token: this.getToken(user.userId),
      userId: user.id,
      username: user.username,
      photo_url: user.photo_url,
      balance: user.balance,
    };
  }

  verifyToken({ token }) {
    if (!token) throw new Error(this.errorCodes.missingToken);
    try {
      const { userId } = jwt.verify(token, process.env.JWT_SECRET || 'secret');
      return { userId };
    } catch (error) {
      throw new Error(this.errorCodes.invalidToken);
    }
  }
}
