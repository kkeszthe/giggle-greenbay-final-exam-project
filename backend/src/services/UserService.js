export class UserService {
  constructor(userRepo, errorCodes) {
    this.user = userRepo;
    this.errorCodes = errorCodes;
  }

  validateParams({ username, password }) {
    if (!username && !password)
      throw new Error(this.errorCodes.missingUsernameAndPassword);
    if (!username) throw new Error(this.errorCodes.missingUsername);
    if (!password) throw new Error(this.errorCodes.missingPassword);
    if (password.length < 8) throw new Error(this.errorCodes.invalidPassword);
  }

  validateURL(string) {
    var pattern = new RegExp(
      '^(http://www.|https://www.|http://|https://)?[a-z0-9]+([-.]{1}[a-z0-9]+)*.[a-z]{2,5}(:[0-9]{1,5})?(/.*)?'
    );
    return pattern.test(string);
  }

  async add({ username, password, photo_url }) {
    this.validateParams({ username, password });
    if (photo_url || this.validateURL(photo_url))
      throw new Error(this.errorCodes.invalidURL);
    const defaultBalance = 0;
    const userId = (
      await this.user.add({
        username,
        password,
        photo_url,
        balance: defaultBalance,
      })
    ).insertId;

    return {
      id: userId,
      username: username,
    };
  }

  async getByUsernameAndPassword({ username, password }) {
    if (!username) throw new Error(this.errorCodes.missingUsername);
    if (!password) throw new Error(this.errorCodes.missingPassword);
    return (await this.user.getAuthentication({ username, password }))[0];
  }

  async getById({ userId }) {
    return (await this.user.getById({ userId }))[0];
  }

  async updateBalance({ userId, amount }) {
    if (!amount) throw new Error(this.errorCodes.missingParam);
    const balance = ((await this.getById({ userId })).balance += parseInt(
      amount
    ));
    await this.user.updateBalance({ userId, balance });
    return await this.getById({ userId });
  }
}
