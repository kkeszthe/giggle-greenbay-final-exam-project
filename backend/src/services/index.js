export * from './UserService';
export * from './SessionService';
export * from './ProductService';
