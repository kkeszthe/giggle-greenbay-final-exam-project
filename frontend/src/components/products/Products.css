.products {
  display: grid;
  grid-template-columns: 1fr 1fr;
  align-items: center;
  gap: 10px;
}

.products .icons {
  display: flex;
  height: 10em;
  width: 20em;
  background-color: white;
  padding: 10px;
}
.products img {
  height: 10em;
  width: 10em;
}
