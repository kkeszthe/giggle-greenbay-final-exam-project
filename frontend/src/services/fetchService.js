import { env } from '../env';

export const generalFetch = async (endpoint, { method, token, body }) => {
  const result = await fetch(`${env.BACKEND_URL}/api/${endpoint}`, {
    method,
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
      token,
    },
    body: JSON.stringify(body),
  });
  return result.json();
};
