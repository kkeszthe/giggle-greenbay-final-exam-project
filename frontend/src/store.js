import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';

import { userReducer } from './reducers/userReducer';
import { productsReducer } from './reducers/productsReducer';
import { tokenReducer } from './reducers/tokenReducer';
import { errorReducer } from './reducers/errorReducer';

function root(state = {}, action) {
  return {
    user: userReducer(state.user, action),
    products: productsReducer(state.troops, action),
    token: tokenReducer(state.token, action),
    error: errorReducer(state.error, action),
  };
}

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
const store = createStore(
  root,
  /* preloadedState, */ composeEnhancers(applyMiddleware(thunk))
);

export { store };
