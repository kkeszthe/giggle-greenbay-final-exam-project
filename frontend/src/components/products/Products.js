import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import { useHistory } from 'react-router-dom';
import CircularProgress from '@material-ui/core/CircularProgress';
import Typography from '@material-ui/core/Typography';

import './Products.css';
import { getProductsAction } from '../../actions/ProductsActions';

function Products({ products, getProducts }) {
  const history = useHistory();

  useEffect(() => {
    getProducts();
  }, [getProducts]);

  function gotToProduct(productId) {
    history.push(`/products/${productId}`);
  }
  if (products.length === 0) {
    return <CircularProgress color="inherit" />;
  }
  return (
    <section className="products">
      {products.map(product => (
        <div
          className="icons"
          key={product.id}
          onClick={() => gotToProduct(product.id)}
        >
          <img
            src={product.product_photo}
            alt={product.product_name}
            className="image"
          ></img>
          <div>
            <Typography variant="body1">{product.product_name}</Typography>
            <Typography variant="body1">{product.price} GBD</Typography>
          </div>
        </div>
      ))}
    </section>
  );
}

const mapStateToProps = ({ products }) => {
  return { products };
};

const mapDispatchToProps = dispatch => {
  return {
    getProducts: () => {
      return dispatch(getProductsAction());
    },
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(Products);
