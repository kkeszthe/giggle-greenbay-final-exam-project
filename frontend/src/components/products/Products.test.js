import React from 'react';
import { Provider } from 'react-redux';
import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import configureStore from 'redux-mock-store';

import Products from './Products';
import { store } from '../../store';
jest.mock('../.././services/fetchService');
import { generalFetch } from '../../services/fetchService';

generalFetch.mockImplementation(() =>
  Promise.resolve([
    {
      id: 15,
      product_name: 'practice',
      description: 'dfghjkéá-l.kg,jhvmnbdvésliretzhulkejsd',
      product_photo:
        'https://doopshop.hu/wp-content/uploads/2019/05/SPIGEN-AIRSKIN-IPHONE-XR-BLACK.jpg',
      price: 12,
      seller_id: 11,
      buyer_id: 11,
      seller_name: 'sal8',
      seller_photo:
        'https://www.nordmet.se/wp-content/uploads/2017/11/Siluett-med-spelarprofilbild-2.jpg',
    },
    {
      id: 15,
      product_name: 'practice',
      description: 'dfghjkéá-l.kg,jhvmnbdvésliretzhulkejsd',
      product_photo:
        'https://doopshop.hu/wp-content/uploads/2019/05/SPIGEN-AIRSKIN-IPHONE-XR-BLACK.jpg',
      price: 12,
      seller_id: 11,
      buyer_id: 11,
      seller_name: 'sal8',
      seller_photo:
        'https://www.nordmet.se/wp-content/uploads/2017/11/Siluett-med-spelarprofilbild-2.jpg',
    },
  ])
);

const mockStore = configureStore([]);
const mockedStore = mockStore({
  user: { username: 'username' },
  products: [
    {
      id: 15,
      product_name: 'practice',
      description: 'dfghjkéá-l.kg,jhvmnbdvésliretzhulkejsd',
      product_photo:
        'https://doopshop.hu/wp-content/uploads/2019/05/SPIGEN-AIRSKIN-IPHONE-XR-BLACK.jpg',
      price: 12,
      seller_id: 11,
      buyer_id: 11,
      seller_name: 'sal8',
      seller_photo:
        'https://www.nordmet.se/wp-content/uploads/2017/11/Siluett-med-spelarprofilbild-2.jpg',
    },
  ],
});
mockedStore.dispatch = jest.fn();

let container;
beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

it('should dispatch an action when rendered', async () => {
  await act(async () => {
    render(
      <Provider store={mockedStore}>
        <Products />
      </Provider>,
      container
    );
  });
  expect(mockedStore.dispatch).toHaveBeenCalledTimes(1);
});

it('matches snapeshot', async () => {
  Storage.prototype.getItem = jest.fn(key => {
    return 'dummy';
  });

  await act(async () => {
    render(
      <Provider store={mockedStore}>
        <Products />
      </Provider>,
      container
    );
  });

  expect(container.innerHTML).toMatchSnapshot();
});
