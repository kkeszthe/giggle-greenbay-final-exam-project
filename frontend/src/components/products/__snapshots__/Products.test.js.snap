// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapeshot 1`] = `"<section class=\\"products\\"><div class=\\"icons\\"><img src=\\"https://doopshop.hu/wp-content/uploads/2019/05/SPIGEN-AIRSKIN-IPHONE-XR-BLACK.jpg\\" alt=\\"practice\\" class=\\"image\\"><div><p class=\\"MuiTypography-root MuiTypography-body1\\">practice</p><p class=\\"MuiTypography-root MuiTypography-body1\\">12 GBD</p></div></div></section>"`;
